src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,248 @@
+#include "mfa_hierarchy.h"
+
+#include <google/protobuf/stubs/casts.h>
+
+#include <Eigen/Dense>
+#include <stan/math/prim/prob.hpp>
+#include <vector>
+
+#include "algorithm_state.pb.h"
+#include "hierarchy_prior.pb.h"
+#include "ls_state.pb.h"
+#include "src/utils/rng.h"
+
+double MFAHierarchy::like_lpdf(const Eigen::RowVectorXd & datum) const {
+
+    return stan::math::normal_lpdf(
+        datum, state.mu + state.Lambda*state.lambda.transpose(), state.psi.cwiseSqrt());
+}
+
+MFA::State MFAHierarchy::draw(const MFA::Hyperparams& params) {
+    auto& rng = bayesmix::Rng::Instance().get();
+    MFA::State out;
+
+    for (size_t j = 0; j < p; j++) {
+        out.mu[j] = stan::math::normal_prec_rng(params.mutilde[j], 1 / sqrt(params.phi), rng);
+        out.psi[j] = stan::math::inv_gamma_rng(params.alpha0, params.beta[j], rng);
+        for (size_t i = 0; i < q; j++) {
+            out.Lambda[j, i] = stan::math::normal_prec_rng(0, 1, rng);
+        }
+    }
+
+    for (size_t i = 0; i < card; i++) {
+        for (size_t j = 0; j < q; j++) {
+            out.Eta[j, i] = stan::math::normal_prec_rng(0, 1, rng);
+        }
+    }
+
+    return out;
+}
+
+void MFAHierarchy::initialize_state() {
+    state.mu = hypers->mutilde;
+    state.psi = hypers->beta / (hypers->alpha0 + 1.);
+    state.Eta = Eigen::MatrixXd::Zero(card,q);
+    state.Lambda = Eigen::MatrixXd::Zero(p,q);
+}
+
+void MFAHierarchy::initialize_hypers() {
+
+    if (prior->has_fixed_values()) {
+        // Set values
+        hypers->mutilde = bayesmix::to_eigen(prior->fixed_values().mutilde());
+        p = hypers->mutilde.size();
+        hypers->beta = bayesmix::to_eigen(prior->fixed_values().beta());
+        hypers->phi = prior->fixed_values().phi();
+        hypers->alpha0 = prior->fixed_values().alpha0();
+        hypers->q = prior->fixed_values().q();
+
+
+        
+        // Check validity
+        if (p != hypers->beta.rows()) {
+            throw std::invalid_argument(
+                "Hyperparameters dimensions are not consistent");
+        }
+        
+        if (hypers->beta <= 0) {
+            throw std::invalid_argument("Shape parameter must be > 0");
+        }
+        if (hypers->alpha0 <= 0) {
+            throw std::invalid_argument("Scale parameter must be > 0");
+        }
+        if (phi->beta <= 0) {
+            throw std::invalid_argument("Diffusion parameter must be > 0");
+        }
+        if (phi->q <= 0) {
+            throw std::invalid_argument("Number of factors must be > 0");
+        }
+    }
+
+    else {
+        throw std::invalid_argument("Unrecognized hierarchy prior");
+    }
+}
+
+void MFAHierarchy::update_hypers(
+    const std::vector<bayesmix::AlgorithmState::ClusterState>& states) {
+    auto& rng = bayesmix::Rng::Instance().get();
+    if (prior->has_fixed_values()) {
+        return;
+    }
+
+    else {
+        throw std::invalid_argument("Unrecognized hierarchy prior");
+    }
+}
+
```

Comment:
We store data in a deque, is this the correct procedure to remove data from the deque according to how the function is called in other parts of the code?

Reply:
probably a `deque` is not the best option here, since removing elements from it is pretty expensive. You could consider a `set`, but that would be problematic if (by chance) there are two observations that are identical. Overall I think that an `std::vector` could provide a strong baseline here, later on we might consider having an `std::map` that stores the values and counts (in case of repeated observations)